### ISSUE TITLE
Rule proposal: AX_ARIA_10

### ISSUE THREAD

**Issue by lencioni** [evcohen/eslint-plugin-jsx-a11y#17]
https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_10

This element has an unsupported ARIA attribute

Many ARIA attributes (states and properties) can only be used on elements with particular roles.

For example `aria-posinset` is only supported by `listitem`, `option`, `menuitemradio`, `radio` and `treeitem`.

Similar to #12

**Comment by evcohen** [comment]
Built out [AX_ARIA_11](https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_11) today. I like all of these proposals. I'm thinking about maybe implementing all of them within the week and publishing v1.0.0

**Comment by evcohen** [comment]
Does this rule mean that role has to be defined for these aria-* properties to be valid? I.E. can you use `aria-posinset` without role being defined?

```jsx
<li aria-posinset="16" />
```

**Comment by lencioni** [comment]
My hunch was that your example should not warn, but I found conflicting answers on stack overflow:

- http://stackoverflow.com/a/11131648/18986
- http://stackoverflow.com/a/11355793/18986

Based on [this information](https://www.w3.org/TR/aria-in-html/#aria-adds-nothing-to-default-semantics-of-most-html-elements), I'd say that `li` has a default role and this rule should not warn. 

Furthermore, it looks like the code for the rule I linked in my OP, [calls a `getRoles` utility method](https://github.com/GoogleChrome/accessibility-developer-tools/blob/e70dd633a25e8a8c659a587f7ce58b923db1eef2/src/audits/UnsupportedAriaAttribute.js#L45) with an argument that [allows implicit semantics to be considered if there is no role attribute](https://github.com/GoogleChrome/accessibility-developer-tools/blob/e70dd633a25e8a8c659a587f7ce58b923db1eef2/src/js/AccessibilityUtils.js#L618-L653).

This seems like enough evidence to me to allow implicit semantics in the eslint rule.

**Comment by evcohen** [comment]
Ah, great. Google's implicit role definition is [here](https://github.com/GoogleChrome/accessibility-developer-tools/blob/master/src/js/Constants.js#L1087). 

So the logic goes

1. Arrive at aria-* attribute node
2. Check the element that contains this attribute for an explicit role, then implicit role.
    - If role is defined, enforce rule accordingly.
    - If role is undefined and there is implicit definition, enforce rule accorindgly.
    - If role is undefined and there is no implicit definition, then **pass or fail?**

I'm thinking in the last case, that we **don't** warn - we can assume an element without an explicit and implicit role can the global properties set (all `aria-*` properties).

**Comment by lencioni** [comment]
I'm not sure about not warning if there is no role and no implicit definition. It seems that if there is no explicit role and no implicit definition, then we would actually want to warn.

**Comment by lencioni** [comment]
Some of the constants in this file might be useful: https://github.com/GoogleChrome/accessibility-developer-tools/blob/e70dd633a25e8a8c659a587f7ce58b923db1eef2/src/js/Constants.js#L1108

It defines the implicit roles for elements, and appears to define lists of allowed roles for each element as well (which might make another good rule here).

**Comment by evcohen** [comment]
Thanks! That's what I've been using as my main reference for outlining attributes across, roles, aria, and DOM elements.

In the third case, if I'm follow [this code](https://github.com/GoogleChrome/accessibility-developer-tools/blob/e70dd633a25e8a8c659a587f7ce58b923db1eef2/src/audits/UnsupportedAriaAttribute.js#L49) correctly, it seems that they are applying the global set of `aria-*` properties to an undefined role.

**Comment by lencioni** [comment]
Ah, interesting. That does seem to be coded that way. However there is that note about it ignoring some problems. I suppose it's not a bad place to start--we can always make it more strict in the future as the plugin is refined.
